### PR TITLE
[Feauture] MoE refactor

### PR DESCRIPTION
**Pull request by Hz188**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [ ] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/modeling/mixtral.py [hunk 1]
```diff
@@ -26,7 +37,7 @@ def setup_ep(self, ep_group: ProcessGroup):
         self.expert_start_idx = self.ep_rank * self.num_experts_per_ep
         held_experts = self.experts[self.expert_start_idx : self.expert_start_idx + self.num_experts_per_ep]
         set_tensors_to_none(self.experts, exclude=set(held_experts))
-        for p in self.experts.parameters():
+        for n, p in self.experts.named_parameters():
```
Comment (ver217) [inline colossalai/shardformer/modeling/mixtral.py:40]: Where you use `n`?

File: colossalai/shardformer/modeling/mixtral.py [hunk 2]
```diff
@@ -1,13 +1,24 @@
...
+from torch.distributed import ProcessGroup
 
 # from colossalai.tensor.moe_tensor.moe_info import MoeParallelInfo
-from torch.distributed import ProcessGroup
-from transformers.models.mixtral.modeling_mixtral import MixtralSparseMoeBlock
+from torch.nn import CrossEntropyLoss
+from transformers.models.mixtral.modeling_mixtral import (
+    MixtralSparseMoeBlock,
+    MoeCausalLMOutputWithPast,
+    _prepare_4d_causal_attention_mask,
```
Comment (ver217) [inline colossalai/shardformer/modeling/mixtral.py:13]: `from transformers.modeling_attn_mask_utils import _prepare_4d_causal_attention_mask`

File: tests/test_moe/test_kernel.py [hunk 3]
```diff
@@ -1,98 +1,104 @@
...
+    dispatch_data_kernel = dispatch_data_kernel.reshape(num_experts, -1, hidden_size)
+    # combine
+    expert_output = dispatch_data_kernel.reshape(-1, hidden_size)
+    ans_kernel = MoeCombine.apply(expert_output, *route_result_list_kernel)
+
+    # no kernel
+    route_result_list_no_kernel = (
+        torch.load(f"{MOE_TENSOR_PATH}/False_2_{data_type}.pt")
+        if MOE_TENSOR_PATH
+        else torch.load(f"False_2_{data_type}.pt")
```
Comment (ver217) [inline tests/test_moe/test_kernel.py:60]: In which case is  `MOE_TENSOR_PATH` None?

File: colossalai/shardformer/modeling/mixtral.py (hunk 2 above)
Comment (Hz188) [reply colossalai/shardformer/modeling/mixtral.py:13]: done

File: tests/test_moe/test_kernel.py (hunk 3 above)
Comment (Hz188) [reply tests/test_moe/test_kernel.py:60]: For user who test it locally, but maybe not need for other developers.

**Comment by Hz188** [comment]
# MoE Test

image

**Review by botbw — APPROVED** [review]

**Pull request merged**